File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/statement/AssignmentStatementParsingFunction.java
package me.itzisonn_.meazy.parser.pasing_function.statement;

import me.itzisonn_.meazy.MeazyMain;
import me.itzisonn_.meazy.text.TextKt;
import me.itzisonn_.meazy.lexer.TokenTypeSets;
import me.itzisonn_.meazy.lexer.TokenTypes;
import me.itzisonn_.meazy.parser.ParsingContext;
import me.itzisonn_.meazy.lexer.Token;
import me.itzisonn_.meazy.parser.Parser;
import me.itzisonn_.meazy.parser.UnexpectedTokenException;
import me.itzisonn_.meazy.parser.ast.expression.Expression;
import me.itzisonn_.meazy.parser.operator.OperatorType;
import me.itzisonn_.meazy.parser.ast.statement.AssignmentStatement;
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression;
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

@NullMarked
public class AssignmentStatementParsingFunction extends AbstractParsingFunction<AssignmentStatement> {
    public AssignmentStatementParsingFunction() {
        super("assignment_statement");
    }

    @Override
    public AssignmentStatement parse(ParsingContext context, @Nullable Object... extra) {
        Parser parser = context.getParser();
        Expression left = parser.parse(MeazyMain.getDefaultIdentifier("expression"), Expression.class);

        if (parser.getCurrent().getType().equals(TokenTypes.ASSIGN())) {
            parser.next();
            Expression value = parser.parse(MeazyMain.getDefaultIdentifier("expression"), Expression.class);
            return new AssignmentStatement(left, value);
        }
        else if (TokenTypeSets.OPERATOR_ASSIGN().contains(parser.getCurrent().getType())) {
            Token token = parser.getCurrentAndNext();

            Expression value = new OperatorExpression(
                    left,
                    parser.parse(MeazyMain.getDefaultIdentifier("expression"), Expression.class),
                    token.getValue().replaceAll("=$", ""), OperatorType.INFIX
            );

            return new AssignmentStatement(left, value);
        }

        throw new UnexpectedTokenException(parser.getCurrent().getLine(), TextKt.literal("Expected assign operators")); //TODO
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/pasing_function/statement/AssignmentStatementParsingFunction.kt
package me.itzisonn_.meazy.parser.pasing_function.statement

import me.itzisonn_.meazy.lexer.TokenTypeSets.operatorAssign
import me.itzisonn_.meazy.lexer.TokenTypes.assign
import me.itzisonn_.meazy.parser.ParsingContext
import me.itzisonn_.meazy.parser.UnexpectedTokenException
import me.itzisonn_.meazy.parser.ast.expression.Expression
import me.itzisonn_.meazy.parser.ast.expression.OperatorExpression
import me.itzisonn_.meazy.parser.ast.statement.AssignmentStatement
import me.itzisonn_.meazy.parser.operator.OperatorType
import me.itzisonn_.meazy.parser.pasing_function.AbstractParsingFunction
import me.itzisonn_.meazy.parser.pasing_function.expression.ExpressionParsingFunction
import me.itzisonn_.meazy.text.translatable

object AssignmentStatementParsingFunction : AbstractParsingFunction<AssignmentStatement>("assignment_statement") {
    override fun parse(context: ParsingContext, vararg extra: Any?): AssignmentStatement {
        val parser = context.parser
        val left = parser.parse(ExpressionParsingFunction)

        if (parser.current.type == assign) {
            parser.next()
            val value = parser.parse(ExpressionParsingFunction)
            return AssignmentStatement(left, value)
        }
        else if (parser.current.type in operatorAssign) {
            val token = parser.consume()

            val value: Expression = OperatorExpression(
                left,
                parser.parse(ExpressionParsingFunction),
                token.value.replace("=$".toRegex(), ""), OperatorType.INFIX
            )

            return AssignmentStatement(left, value)
        }

        throw UnexpectedTokenException(parser.current.line, translatable("meazy:parser.expected.separator_statement", "assign", "assignment"))
    }
}
